flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -215,6 +224,11 @@ private void cancelTask(boolean interrupt) {
     }
 
     private void interruptSourceThread(boolean interrupt) {
+        // Nothing need to do if the source is finished on restore
+        if (operatorChain != null && operatorChain.isFinishedOnRestore()) {
```

Comment:
Here we should judge if `operatorChain` is null since `cancel` might happen before the task is successfully initialized. 

Although the `cancel` might be called from different thread other than the main thread, there might be small probability the `operatorChain` is set but not detected, but this should also not cause problem in considering the following process. Thus for now we would not change `operatorChain` to `volatile`, which is also similar to the current `mainOperator` variable.

Reply:
Thanks for the explanation